Modules/hmacmodule.c

```diff
@@ -0,0 +1,1713 @@
+
+    // HMAC HACL* internal state.
+    HACL_HMAC_state *state;
+} HMACObject;
+
+#define HMACObject_CAST(op) ((HMACObject *)(op))
+
+// --- HMAC module clinic configuration ---------------------------------------
+
+/*[clinic input]
+module _hmac
+class _hmac.HMAC "HMACObject *" "clinic_state()->hmac_type"
+[clinic start generated code]*/
+/*[clinic end generated code: output=da39a3ee5e6b4b0d input=c8bab73fde49ba8a]*/
+
+#define clinic_state()  (get_hmacmodule_state_by_cls(Py_TYPE(self)))
+#include "clinic/hmacmodule.c.h"
+#undef clinic_state
+
+// --- Helpers ----------------------------------------------------------------
+//
+// The helpers have the following naming conventions:
+//
+// - Helpers with the "_hacl" prefix are thin wrappers around HACL* functions.
+//   Buffer lengths given as inputs should fit on 32-bit integers.
+//
+// - Helpers with the "hmac_" prefix act on HMAC objects and accept buffers
+//   whose length fits on 32-bit or 64-bit integers (depending on the host
+//   machine).
+
+/*
+ * Assert that a HMAC hash kind is a static kind.
+ *
+ * A "static" kind is specified in the 'py_hmac_static_hinfo'
+ * table and is always independent of the host CPUID features.
+ */
+#ifndef NDEBUG
+static void
+assert_is_static_hmac_hash_kind(HMAC_Hash_Kind kind)
+{
+    switch (kind) {
+        case Py_hmac_kind_hash_unknown: {
+            Py_FatalError("HMAC hash kind must be a known kind");
+            return;
+        }
+        case Py_hmac_kind_hmac_vectorized_blake2s_32:
+        case Py_hmac_kind_hmac_vectorized_blake2b_32: {
+            Py_FatalError("HMAC hash kind must not be a vectorized kind");
+            return;
+        }
+        default:
+            return;
+    }
+}
+#else
+static inline void
+assert_is_static_hmac_hash_kind(HMAC_Hash_Kind Py_UNUSED(kind)) {}
+#endif
+
+/*
+ * Convert a HMAC hash static kind into a runtime kind.
+ *
+ * A "runtime" kind is derived from a static kind and depends
+ * on the host CPUID features. In particular, this is the kind
+ * that a HMAC object internally stores.
+ */
+static HMAC_Hash_Kind
+narrow_hmac_hash_kind(hmacmodule_state *state, HMAC_Hash_Kind kind)
+{
+    switch (kind) {
+        case Py_hmac_kind_hmac_blake2s_32: {
+#if HACL_CAN_COMPILE_SIMD128
+            if (state->can_run_simd128) {
+                return Py_hmac_kind_hmac_vectorized_blake2s_32;
+            }
+#endif
+            return kind;
+        }
+        case Py_hmac_kind_hmac_blake2b_32: {
+#if HACL_CAN_COMPILE_SIMD256
+            if (state->can_run_simd256) {
+                return Py_hmac_kind_hmac_vectorized_blake2b_32;
+            }
+#endif
+            return kind;
+        }
+        default:
+            return kind;
+    }
+}
+
+/*
+ * Handle the HACL* exit code.
+ *
+ * If 'code' represents a successful operation, this returns 0.
+ * Otherwise, this sets an appropriate exception and returns -1.
+ */
+static int
+_hacl_convert_errno(hacl_errno_t code, PyObject *algorithm)
```

Comment:
This function is needed for HMAC because:

- malloc may return OutOfMemory, InvalidAlgorithm (e.g. requesting Blake2b_256 on an ARM machine), or Success
- reset may return InvalidLength (if trying to reset the state with a key of different length, this is not supported), or Success
- update: MaximumLengthExceeded or Success
- digest: OutOfMemory or Success
- copy: may return NULL (indicates out of memory)

See https://github.com/python/cpython/pull/130960#issuecomment-2715631704.